airflow/models/dag.py

```diff
@@ -1207,6 +1207,18 @@ def tasks(self, val):
     def task_ids(self) -> list[str]:
         return list(self.task_dict.keys())
 
+    @property
+    def teardowns(self):
+        return [task for task in self.tasks if getattr(task, "_is_teardown", None)]
+
+    @property
+    def tasks_upstream_of_teardowns(self):
+        upstream_tasks = [t.upstream_list for t in self.teardowns]
+        flatten_tasks = [
+            val for sublist in upstream_tasks for val in (sublist if isinstance(sublist, list) else [sublist])
+        ]
+        return [val for val in flatten_tasks if not getattr(val, "_is_teardown", None)]
```

Comment:
nit: We might be able to refactor this a bit to do one let loop through, without impacting readability much.

```suggestion
        upstream_tasks = [t.upstream_list if isinstance(t.upstream_list, list) else [t.upstream_list]) t.upstream_list for t in self.teardowns]
        return [
            val for sublist in upstream_tasks for val in sublist if not getattr(val, "_is_teardown", None)
        ]
```

Reply:
`upstream_list` can't be a single task, right? It always returns a list of tasks?

https://github.com/apache/airflow/blob/366ae51e21dfb70da85962ca753e8cba8c4d3ec8/airflow/models/taskmixin.py#L254-L258

That'd simplify it even further.